com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -104,29 +116,166 @@ public int AddEntry(ulong networkObjectId, int behaviourIndex, int variableIndex
+            writer.WriteUInt64(entry.Key.NetworkObjectId);
+            writer.WriteUInt16(entry.Key.BehaviourIndex);
+            writer.WriteUInt16(entry.Key.VariableIndex);
+            writer.WriteUInt16(entry.Key.TickWritten);
+            writer.WriteUInt16(entry.Position);
+            writer.WriteUInt16(entry.Length);
+        }
+
+        /// <summary>
+        /// Read a received Entry
+        /// Must match WriteEntry
+        /// </summary>
+        /// <param name="reader">The readed to read the entry from</param>
+        internal Entry ReadEntry(NetworkReader reader)
+        {
+            Entry entry;
+            entry.Key.NetworkObjectId = reader.ReadUInt64();
+            entry.Key.BehaviourIndex = reader.ReadUInt16();
+            entry.Key.VariableIndex = reader.ReadUInt16();
+            entry.Key.TickWritten = reader.ReadUInt16();
+            entry.Position = reader.ReadUInt16();
+            entry.Length = reader.ReadUInt16();
+            entry.Fresh = false;
+
+            return entry;
+        }
+
         /// <summary>
         /// Allocate memory from the buffer for the Entry and update it to point to the right location
         /// </summary>
         /// <param name="entry">The entry to allocate for</param>
         /// <param name="size">The need size in bytes</param>
-        public void AllocateEntry(ref Entry entry, long size)
+        public void AllocateEntry(ref Entry entry, int index, int size)
         {
             // todo --M1--
             // this will change once we start reusing the snapshot buffer memory
             // todo: deal with free space
             // todo: deal with full buffer
 
-            entry.Position = (ushort)FreeMemoryPosition;
+            int pos;
+
+            if (entry.Length > 0)
+            {
+                Allocator.Deallocate(index);
+            }
+
+            bool ret = Allocator.Allocate(index, size, out pos);
+
+            if (!ret)
+            {
+                //todo: error handling
+            }
+
+            entry.Position = (ushort)pos;
             entry.Length = (ushort)size;
-            FreeMemoryPosition += (int)size;
+        }
+
+        /// <summary>
+        /// Read the buffer part of a snapshot
+        /// Must match WriteBuffer
+        /// The stream is actually a memory stream and we seek to each variable position as we deserialize them
+        /// </summary>
+        /// <param name="reader">The NetworkReader to read our buffer of variables from</param>
+        /// <param name="snapshotStream">The stream to read our buffer of variables from</param>
+        internal void ReadBuffer(NetworkReader reader, Stream snapshotStream)
+        {
+            int snapshotSize = reader.ReadUInt16();
+
+            snapshotStream.Read(Buffer, 0, snapshotSize);
+
+            for (var i = 0; i < LastEntry; i++)
+            {
+                if (Entries[i].Fresh && Entries[i].Key.TickWritten > 0)
```

Comment:
Do we have to do a linear search from 0 until we find `Fresh` ones?  Like could we just have an index into the first Fresh one rather than scan down to it?

For some reason I figured `Entries` would be a circular buffer, but maybe instead when we get a new entry we just find the first un-fresh slot and stick it there.

Reply:
Yes, good points. Maybe I was expecting code to prematurely go into `develop` and make in in stages.

Once we have acknowledgements, variables will be removed from the main snapshots. When the last variable is removed `LastEntry` will decrease back. This PR was only meant for the M1b milestone and getting the scale-demo to pass.